2 countries demo

koriaf opened the pull request:
sorry to break pie.py, I'm not sure I'll be able to fix it today - so just sending it for review. may be someone else fixes that script :-)

I haven't tied chambers/importers/others app to it for the same reason

koriaf committed:
feat(channel): Some minor changes to better local channel dev

koriaf committed:
feat(demo-docker): 2 setups - AU and CN - with readme

* Docker-compose files creating 2 separate networks (still able to talk to each other)
* readme updated about how to start it
* broke the pie.py script because old docker-compose.yml has been deleted
* removed binding API ports to the host container - if anyone uses it please uncomment
* run flask services on 80 port because why not

adamkerz commented:
A two countries demo goes beyond a Node, it also requires 1 or more channels to be configured. I think I wrote about this a while ago, but only generally. It would be good if we could have a single docker-compose file that starts up services for a node, but that is configurable enough that we can just set environment variables to start two nodes for two separate countries. Eg.

```
PROJECT_NAME=igl_node_au # for docker-compose separation
COUNTRY_CODE=AU
STARTING_PORT_NUMBER=1234 # and each service within the node could use STARTING_PORT_NUMBER+1 or 2 or 5 or whatever to maintain separation
# ... I'm sure other vars are required
pie intergov.start_node
```

Then configure the routing table:

```
ROUTING_TABLE_FILE=routing_table_au.json
pie intergov.configure_routing_table
```

Rinse and repeat for other nodes

Then start up whatever channels are needed, in a similar fashion:
```
SHARED_DB=postgres:blah
API_PORT_NUMBER=8765
pie shared_db_channel_endpoint.start
```

And start the instance of the shared db:
```
DB_PORT_NUMBER=9999
pie shared_db_channel.start
```

The shared_db tasks would live in the shared_db_channel repo.

monkeypants commented:
@koriaf, do you understand the Chambers API stuff well enough to do this, or do you want some help from @adamkerz / @magicjohnson ?

koriaf commented:
This single docker-compose will be horribly big; but yes, it requires channel, chambers and importer app at least to be configured well.
I just tend to use approach "use this to start intergov if you need nothing else, then start chambers app of you need chambers app, then start channels and configured them if you need them and so on". It will keep services more or less separate but still linked and working together, and will share purpose of demo files to development as well (I don't need channels and chambers to work on message api, and I don't need chambers to work on channel api integration).

There will be README updated reflecting that; and it definitely won't be one-click process (because it looks pointless to me), but should be quick (5-8 minutes including reading excluding containers build time)

Regarding the chambers app integration: I don't think anybody except me will do it quick (given it's my project). It's the next step (I already have channels working) - just need to approve the whole approach.

adamkerz commented:
No, I'm not suggesting a single docker-compose file. Four or more: intergov, importer app, chambers app, shared_db channel, and more for other channel implementations.

But from that, we could write scripts (pie_tasks) that have default values to allow two or three nodes, several channels and several apps to all be configured and spun up to simulate a real world situation to test end to end stuff - in just a few commands.

And if you want to do something more complex, you can override the defaults with environment variables. Eg. if the above provides two nodes, you can set some env vars and spin up a third node, channel, and required apps using the same commands.

koriaf commented:
and are you okay with single docker-compose file per country?

adamkerz commented:
Not really - what is different between each country? Just port numbers, the country code and the routing table. And auth config, and keys to connect to the channel and... OK, so there's a bunch of things. But I see a lot of duplication if we have two docker-compose files instead of one that can be configured to act as one country or another. Convince me otherwise :-)

koriaf commented:
I'll try to do it with the single one

koriaf committed:
feat(demo): Merge 2 docker-compose files per country to a single one

koriaf committed:
feat(demo): Add docker-host and update network names to lowercase

koriaf commented:
okay, now I can confirm that the chambers app and the shared DB workers are linked to the current setup, and it works fine for me so far - I've started AU and SG setups, and linked the chambers app to the AU one, and it sends certificate either to the cloud-based shared channel we have (AU, SG countries) or to the locally started shared-db on the host container 7500 port (France). There will be more changes to make it look better. I think it's 50% through.

adamkerz committed:
Getting working on Windows and with pie_tasks

adamkerz commented on demo.yml:
```diff
@@ -50,36 +42,46 @@ x-services:
-    ports: ['9324:9324']
     volumes: ['./tests/local_elasticmq.conf:/opt/elasticmq.conf']
     logging:
-      # because it's noisy
```
Was this the reason for logging being disabled for Postgres? It made it painful to figure out why postgres wouldn't start for me. Changing the line and restarting the postgres service didn't work - I had to down and then up the whole stack for it to re-read the change :-(

koriaf replied:
:-(
I hope you won't need these logs when it's working well :-)

adamkerz commented on demo.yml:
```diff
@@ -50,36 +42,46 @@ x-services:
-    # ports: ['5432:5432']
     networks:
       - internal
+    volumes:
```
I turned logging back on and, as mentioned in a comment above, local path volumes on Windows don't get proper permissions so named volumes is the only way to get postgres to run but not lose the contents of the volume if you stop and start the services.

koriaf replied:
okay, sorry. I'm fine with the docker volumes as well - it was just useful for me to have it persistent even when all other volumes are deleted (on some stage)

koriaf replied:
are you okay with the minio on windows? It's very comfortable for me to have that path shared with the host machine because I can empty and check the repos content easily using my  file browser, but if you hit the same issue then please do it the same - thus removing any need in `var*` directories.

adamkerz committed:
Bit of a hack, to be fixed better when we define our different methods of testing better

adamkerz committed:
Damn, how did I undo that before committing?!

koriaf commented on Jenkinsfile:
```diff
@@ -53,7 +53,7 @@ pipeline {
                             sh '''#!/bin/bash
                                 export COMPOSE_PROJECT_NAME=au
-                                cp demo-au.env demo-au.env
+                                cp demo-au.env demo-au-local.env
```
please note these files are different. demo-au-local is fine to be empty by default (mine contains IGL_OAUTH_CLIENT_ID and other things to work with the shared DB channel deployed by Michael)

so it's fine to create empty demo-*-local.env file; but copying works fine for the same reason - just re-defines variables in their same values

adamkerz replied:
Don't we need to at least define `IGL_COUNTRY=AU`? If not now, I suspect that it will be mandatory for a node to have that explicitly defined.

Anyway, as we make clearer the distinction between just spinning up one or more nodes and running a specific set of nodes and channels for the purpose of running integration tests, I expect this code to change.

koriaf replied:
![Screenshot from 2020-05-21 09-11-51](https://user-images.githubusercontent.com/131850/82533229-2723e780-9b43-11ea-91e8-d4e2eb6ad195.png)

look, all 3 files are used. so demo-au.env contains exactly that, and demo-au-local contains whatever you need for you locally.

for integration tests insinde the single country we just start up any of it (au, cn). integration tests taking 2 or more countries will require starting 2.

adamkerz replied:
Ahhhh, sorry, yes, I see! That makes much more sense - I approve!

koriaf committed:
feat(demo): Move routing table to the env variable

koriaf commented:
okay, I've moved routing table to some env variable and updated the readme; now I think this PR is atomic enough to be merged - but commits should  be squashed on merge to loose that pointless history

koriaf commented:
and it won't be deployed well untill @roleyfoley updates the `IGL_MCHR_ROUTING_TABLE` variable to the corresponding value (setting up AU SG and CN channels for example)
```
[
    {
        "Name": "shared db channel to Australia",
        "Jurisdiction": "AU",
        "Predicate": "UN.CEFACT.",
        "ChannelUrl": "https://sharedchannel.services.devnet.trustbridge.io/",
        "ChannelAuth": "Cognito/JWT",
    },
    {
        "Name": "shared db channel to SG",
        "Jurisdiction": "SG",
        "Predicate": "UN.CEFACT.",
        "ChannelUrl": "https://sharedchannel.services.devnet.trustbridge.io/",
        "ChannelAuth": "Cognito/JWT",
    },
    {
        "Name": "shared db channel to CN",
        "Jurisdiction": "CN",
        "Predicate": "UN.CEFACT.",
        "ChannelUrl": "https://sharedchannel.services.devnet.trustbridge.io/",
        "ChannelAuth": "Cognito/JWT",
    }
]
```

koriaf commented:
I'm able to send messages from the Chambers App to the local AU intergov (because it's looking at it), which when routes it either to shared cloud channel or my local one. Then chambers app receives the notification back about message changing it's status to "Accepted" (if everything is fine) or "Rejeccted" (if you are sending the message to some country which has no channels configured).

shared channel DB contains rows like this:
```
 2020-05-20 13:23:59.441649 |  1 | {"sender": "AU", "receiver": "FR", "subject": "AU.Chambers_AU-AU.coo9913f4", "obj":
 "Qma3L31y1cDqnFisCsJYDaE6Jz3kshNA5DbpZrAPUntm9w", "predicate": "UN.CEFACT.Trade.CertificateOfOrigin.created"} | confi
rmed
 2020-05-21 07:23:51.00956  |  2 | {"sender": "AU", "receiver": "FR", "subject": "AU.Chambers_AU-AU.coo9913f4", "obj":
 "Qma3L31y1cDqnFisCsJYDaE6Jz3kshNA5DbpZrAPUntm9w", "predicate": "UN.CEFACT.Trade.CertificateOfOrigin.created"} | confi
rmed
```

chambers app demo scripts were updated and pushed to it's master, shared DB channel needs to change, intergov devops needs a new variable.

roleyfoley commented:
> and it won't be deployed well untill @roleyfoley updates the `IGL_MCHR_ROUTING_TABLE` variable to the corresponding value (setting up AU SG and CN channels for example)

Added in ready for next deployment. 

Also moved from `intergov/processors/blockchain_router/__init__.py` to `intergov/processors/multichannel_router/__init__.py` feel free to remove the symlink when ready

monkeypants commented:
Thanks Roley, is this still WIP?

adamkerz committed:
Removing symlink

adamkerz commented:
I'll merge once my pushed change passes tests

monkeypants commented on DEVELOPMENT.rst:
```diff
@@ -18,28 +16,44 @@ TODO: create readthedocs site and point to it.
 Quickstart
 ----------
 
+Intergov is tighly coupled with the countries, so if you want to start it locally
```
At some point we will need to do a search and replace of "countries" with "jurisdictions", including environment variable names...

monkeypants commented on demo-default.env:
```diff
@@ -51,3 +50,42 @@ IGL_TEST_DBNAME=postgres_test
 IGL_ALLOW_UNSAFE_REPO_CLEAR=True
+
+
+IGL_MCHR_ROUTING_TABLE=[{ "Name": "shared db channel to Australia", "Jurisdiction": "AU", "Predicate": "UN.CEFACT.", "ChannelUrl": "https://sharedchannel.services.devnet.trustbridge.io/", "ChannelAuth": "Cognito/JWT" }, { "Name": "shared db channel to Singapore", "Jurisdiction": "SG", "Predicate": "UN.CEFACT.", "ChannelUrl": "https://sharedchannel.services.devnet.trustbridge.io/", "ChannelAuth":  ...
```
Note: @adamkerz; at some point some of this will need to become dynamic (call channel API to find out what predicates/jurisdictions the channel supports). I *think* the only static/configured part of the routing table should be the precedence-ordered list of channels.

koriaf replied:
yes, I think urls + channel type (always channel API so far but who knows) and auth variables will be the best choice

adamkerz replied:
Of course, @monkeypants, I've already noted that the routing table will be changing to db table of some form as soon as we have it working.

adamkerz commented:
All tests pass locally for me - @koriaf do you have any insight into why they are failing in CI? Something about `MinioRepo._unsafe_is_empty_for_test` - what does "unsafe" refer to? I've seen it a few times and just glossed over it...

koriaf commented:
it means that integration tests empty buckets or delete things, which means that if they are accidentally run on prod environment it will be sad. I think we may disable these asserts, ideally changing the tests so they don't expect empty repos/buckets and work fine with thatever information is there

koriaf commented:
I'm fine with merging it if it doesn't break the build. Unittests must pass but integration ones are too brittle to be reliable.

monkeypants commented:
Created tickets #192 and #193 to keep those TODOs in the collective memory.

Are those integration test failures caused by concurrnet tests sharing the same backends? If so, maybe we need to re-architect them so they have some test-instance namespace on backend paths?

koriaf commented:
> sharing the same backends?

no, I think integration tests are started only for the single country. but it's an interesting note - after we migrated to docker-volumes we probably should name them with the country prefix as well...

monkeypants commented:
So what do we do about these broken integration tests? Merge and raise an issue, or fix then merge?

koriaf commented:
given no developer can reproduce it quickly - I advise just ignore and make sure they don't stop the build.

monkeypants commented:
so should I squash/merge, then see if it works?

koriaf commented:
yes, let's try just to squash it. commit message will be ugly anyway
